packages/nextjs/app/_components/PriceGraph.tsx

```diff
@@ -1,64 +1,177 @@
+    isSavingsRateLoading ||
+    !priceEvents ||
+    !borrowRateUpdatedEvents ||
+    !savingsRateUpdatedEvents;
+
+  const combinedEvents = [
+    ...(priceEvents || []),
+    ...(borrowRateUpdatedEvents || []),
+    ...(savingsRateUpdatedEvents || []),
+  ];
+  const sortedEvents = combinedEvents.sort((a, b) => Number(a.blockNumber - b.blockNumber));
+
+  type DataPoint = {
+    blockNumber: number;
+    price: number;
+    borrowRate: number;
+    savingsRate: number;
+  };
+
+  const priceData = sortedEvents.reduce<DataPoint[]>((acc, event, idx) => {
+    const price = event.eventName === "PriceUpdated" ? 1 / (Number(formatEther(event.args.price || 0n)) / 1800) : 0;
+    const borrowRate = event.eventName === "BorrowRateUpdated" ? Number(event.args.newRate || 0n) / 100 : 0;
+    const savingsRate = event.eventName === "SavingsRateUpdated" ? Number(event.args.newRate || 0n) / 100 : 0;
+
+    const prevPrice = acc[idx - 1]?.price || 1;
+    const prevBorrowRate = acc[idx - 1]?.borrowRate || 0;
+    const prevSavingsRate = acc[idx - 1]?.savingsRate || 0;
 
-  const priceInitial = [{ blockNumber: 0, price: 1 }];
+    return [
+      ...acc,
+      {
+        blockNumber: Number(event.blockNumber) || 0,
+        price: price && Number.isFinite(price) ? price : prevPrice,
+        borrowRate: borrowRate && Number.isFinite(borrowRate) ? borrowRate : prevBorrowRate,
+        savingsRate: savingsRate && Number.isFinite(savingsRate) ? savingsRate : prevSavingsRate,
+      },
+    ];
+  }, []);
 
   return (
     <div className="card bg-base-100 w-full shadow-xl indicator">
-      <TooltipInfo top={3} right={3} infoText="This graph shows the price of the stabletoken over time" />
+      <TooltipInfo
+        top={3}
+        right={3}
+        infoText="This graph displays the stablecoin price (yellow), borrow rate (red), and savings rate (green) over time. Toggle rates visibility using the button."
+      />
       <div className="card-body h-96 w-full">
-        <h2 className="card-title">Price Graph</h2>
-        <ResponsiveContainer width="100%" height="100%">
-          <LineChart width={500} height={300} data={priceEventsArray?.length ? priceEventsArray : priceInitial}>
-            <XAxis
-              domain={["auto", "auto"]}
-              dataKey="blockNumber"
-              stroke={strokeColor}
-              tick={false}
-              label={{ value: "Time (Blocks)", position: "insideBottom", fill: strokeColor }}
-            />
-            <YAxis
-              scale="linear"
-              domain={[(dataMin: number) => dataMin - 0.0001, (dataMax: number) => dataMax + 0.0001]}
-              stroke={strokeColor}
-              tick={{ fill: strokeColor }}
-            />
-            <Tooltip
-              labelFormatter={(value: number) => `Block: ${value}`}
-              formatter={(value: number) => [`$${value.toFixed(6)}`]}
-            />
-            <ReferenceLine y={1.0} stroke="#ff4d4d" strokeDasharray="3 3" />
```

Comment:
I wish we could still have this reference line but I see that the color would conflict with the rate lines. Maybe we should experiment with other colors for the rate lines?